cannon: Encapsulate syscall helpers

mbaxter opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

Encapsulate syscall logic into a number of helper functions  to prep for code-sharing between multi-threaded Cannon (MTCannon) and Cannon. 

Also:
- Move memory accessor helpers from `MIPS.sol` to `MIPSMemory.sol` library
- Fix `fcntl` typos

**Tests**

This is a refactor, existing test coverage should be sufficient.  

**Additional context**

Builds on https://github.com/ethereum-optimism/optimism/pull/10943

**Metadata**

Part of https://github.com/ethereum-optimism/optimism/issues/11639

mbaxter committed:
cannon: Increment MIPS.sol version

mbaxter committed:
cannon: Add syscall args helper

mbaxter committed:
cannon: Extract mmap helper

mbaxter committed:
cannon: Extract sysRead helper

Also:
- Extract solidity memory helpers into new lib.
- Reorganize syscall constants.

mbaxter committed:
cannon: Add NatSpec documentation to solidity syscall helpers

mbaxter committed:
cannon: Extract sysWrite helper

mbaxter committed:
cannon: Use consistent naming convention

mbaxter committed:
cannon: Extract sysFcntl helper, fix typos

mbaxter committed:
cannon: Consolidate comments, fix formatting

mbaxter committed:
cannon: Add helper for setting registers, pc fields

mbaxter committed:
cannon: Reorganize syscall constants

mbaxter committed:
cannon: Fix MIPSMemory import

mbaxter committed:
cannon: Run semver-lock and snapshots

mbaxter commented on packages/contracts-bedrock/src/cannon/libraries/MIPSSyscalls.sol:
```diff
@@ -0,0 +1,266 @@
+            // If the preimage key is a local key, localize it in the context of the caller.
+            if (uint8(_preimageKey[0]) == 1) {
+                _preimageKey = PreimageKeyLib.localize(_preimageKey, _localContext);
+            }
+            (bytes32 dat, uint256 datLen) = _oracle.readPreimage(_preimageKey, _preimageOffset);
+
+            // Transform data for writing to memory
+            // We use assembly for more precise ops, and no var count limit
+            assembly {
+                let alignment := and(_a1, 3) // the read might not start at an aligned address
+                let space := sub(4, alignment) // remaining space in memory word
+                if lt(space, datLen) { datLen := space } // if less space than data, shorten data
+                if lt(_a2, datLen) { datLen := _a2 } // if requested to read less, read less
+                dat := shr(sub(256, mul(datLen, 8)), dat) // right-align data
+                dat := shl(mul(sub(sub(4, datLen), alignment), 8), dat) // position data to insert into memory
+                // word
+                let mask := sub(shl(mul(sub(4, alignment), 8), 1), 1) // mask all bytes after start
+                let suffixMask := sub(shl(mul(sub(sub(4, alignment), datLen), 8), 1), 1) // mask of all bytes
+                // starting from end, maybe none
+                mask := and(mask, not(suffixMask)) // reduce mask to just cover the data we insert
+                mem := or(and(mem, not(mask)), dat) // clear masked part of original memory, and insert data
+            }
+
+            // Write memory back
+            MIPSMemory.writeMem(_a1 & 0xFFffFFfc, 1, mem);
+            newPreimageOffset_ += uint32(datLen);
+            v0_ = uint32(datLen);
+        }
+        // hint response
+        else if (_a0 == FD_HINT_READ) {
+            // Don't read into memory, just say we read it all
+            // The result is ignored anyway
+            v0_ = _a2;
+        } else {
+            v0_ = 0xFFffFFff;
+            v1_ = EBADF;
+        }
+
+        return (v0_, v1_, newPreimageOffset_);
+    }
```
This logic is unchanged.

mbaxter commented on cannon/mipsevm/mips_syscalls.go:
```diff
@@ -0,0 +1,195 @@
+	return v0, v1, newPreimageOffset
+}
+
+func handleSysWrite(a0, a1, a2 uint32, lastHint hexutil.Bytes, preimageKey [32]byte, preimageOffset uint32, oracle PreimageOracle, memory *Memory, memTracker MemTracker, stdOut, stdErr io.Writer) (v0, v1 uint32, newLastHint hexutil.Bytes, newPreimageKey common.Hash, newPreimageOffset uint32) {
```
Some of these function signatures are kind of crazy, but not sure how to avoid it... 🤷‍♀️

Inphi replied:
Yeah. I think it's fine though because there is only one callsite.

mbaxter commented on packages/contracts-bedrock/src/cannon/MIPS.sol:
```diff
@@ -148,277 +139,63 @@ contract MIPS is ISemver {
-            }
-            // brk: Returns a fixed address for the program break at 0x40000000
-            else if (syscall_no == 4045) {
+            if (syscall_no == sys.SYS_MMAP) {
```
Not sure if there was a reason we weren't using constants here before?

Inphi replied:
No good reason. This is better.

mbaxter commented on cannon/mipsevm/mips.go:
```diff
@@ -75,107 +55,22 @@ func (m *InstrumentedState) handleSyscall() error {
 		m.state.ExitCode = uint8(a0)
 		return nil
```
Just a note that I did not make helpers for some of the simpler logic here (sysBrk, sysClone, sysExitGroup).   Leaving sysClone inline definitely makes sense as it is a noop implementation and we'll have a real implementation in MTCannon.

Inphi commented on packages/contracts-bedrock/src/cannon/libraries/MIPSMemory.sol:
```diff
@@ -0,0 +1,121 @@
+library MIPSMemory {
+    /// @notice Reads a 32-bit value from memory.
+    /// @param _addr The address to read from.
+    /// @param _proofIndex The index of the proof in the calldata.
```
For this function and `writeMem`, let's document here the expected proof layout in calldata. Since it's no longer obvious as these routines are in a separate library.
See https://github.com/ethereum-optimism/optimism/pull/10978/files#diff-2543eb93ac56f9072b671119906a62fec2a87591132b4315ef4745b84c8c43cbR110

mbaxter replied:
Thanks for raising this - I realized that there were a few places where these helper functions were making assumptions about the state / memory layout etc.  Reworked these helpers to try to make them agnostic to the state format.  Have a look at the last couple commits and let me know if these updates look good.

Inphi replied:
That's better. It should be more amenable for MT-Cannon later on as we extend the witness representation.
I left a comment on additional calldata checks.

mbaxter committed:
cannon: Explicitly inject proof offsets into helper fns

Proofs offsets may differ between implementations, so these values must
be passed into helper functions rather than calculated internally.

mbaxter committed:
cannon: Remove hard-coded state.memRoot references from MIPSMemory.sol

mbaxter committed:
cannon: Work around stack too deep error

mbaxter committed:
cannon: Rework stack-too-deep fix

mbaxter committed:
cannon: Run semver-lock and snapshots

Inphi commented on packages/contracts-bedrock/src/cannon/libraries/MIPSMemory.sol:
```diff
@@ -0,0 +1,116 @@
+                if and(_addr, 3) { revert(0, 0) }
+
+                // Load the leaf value.
+                let leaf := calldataload(_proofOffset)
```
It's worth asserting here again (and in `writeMem`) that there's sufficient calldata for the proof. (see https://github.com/ethereum-optimism/optimism/pull/10978/files#diff-2543eb93ac56f9072b671119906a62fec2a87591132b4315ef4745b84c8c43cbR112 for example).

Otherwise, misuse of this function will be silent as calling `calldataoad` for offsets beyond actual calldata returns zero.

mbaxter replied:
Good catch - I moved the calldata length validation from `memoryProofOffset` into `readMem` and `writeMem`.

mbaxter committed:
cannon: Validate calldata size directly in readMem, writeMem

mbaxter committed:
cannon: Run semver-lock and snapshots

Inphi reviewed (approved):
Nice!
